service/windows/src/main.rs

```diff
@@ -0,0 +1,312 @@
+                // does not appear to cause any problems.
+                handler_is_shutdown.store(true, Ordering::Release);
+                service_thread.unpark();
+                ServiceControlHandlerResult::NoError
+            }
+            ServiceControl::Interrogate => ServiceControlHandlerResult::NoError,
+            _ => ServiceControlHandlerResult::NotImplemented,
+        }
+    };
+
+    // Register the event handler callback.
+    let status_handle = register(SERVICE_NAME, event_handler)?;
+
+    // Set the status to StartPending while we load the configuration and start the nodes.
+    status_handle.set_service_status(pending_status!(
+        ServiceState::StartPending,
+        0,
+        Duration::from_secs(1)
+    ))?;
+
+    let config = load_config()?;
+
+    // Start the nodes
+    let mut nodes = Vec::new();
+    for node_conf in config.nodes {
+        let node_name = node_conf.name.clone();
+        match node_conf.start() {
+            Ok(node) => {
+                info!("Started node '{}'", &node.node_config.name);
+                nodes.push(node);
+            }
+            Err(e) => error!("Could not start node '{}': {:#}", node_name, e),
+        }
+    }
+
+    // Tell the system that the service is running now
+    status_handle.set_service_status(simple_status!(
+        ServiceState::Running,
+        ServiceControlAccept::STOP | ServiceControlAccept::PRESHUTDOWN
+    ))?;
+
+    // Loop until shutdown is triggered
+    while !is_shutdown.load(Ordering::Acquire) {
+        // Check for any stopped nodes. If a node is stopped, the fact is logged, but we do not
+        // attempt to restart it.
+        nodes.retain_mut(|node| match node.check_exit() {
+            Ok(None) => true,
+            Ok(Some(e)) => {
+                let err_output = node.get_error_output().unwrap_or_default();
+                error!(
+                    "The node '{}' has stopped unexpectedly. ({})\n\nOutput:\n{}",
+                    node.node_config.name, e, err_output
+                );
+                false
+            }
+            Err(e) => {
+                error!(
+                    "The state of node '{}' cannot be determined:\n{}",
+                    node.node_config.name, e
+                );
+                false
+            }
+        });
+        if nodes.is_empty() {
+            warn!("All nodes have stopped. The node runner service will now shut down.");
+            break;
+        }
+        // Wait for a second. This can be woken early by the event handler.
+        thread::park_timeout(std::time::Duration::from_secs(1));
+    }
+
+    // Shut down the nodes.
+    // Nodes are first sent a CTRL+BREAK signal to trigger shutdown gracefully, and are allowed 61
+    // seconds to do so.  Any node that has not shut down in that time will be terminated.
+    let expected_shutdown_duration = Duration::from_secs(61);
```

Comment:
I wanted to pick a number that is long enough that the node should be able to shutdown in that time, and not longer than the usual limits on service shutdown times. The preshutdown time-out is 3 minutes. The Services control panel allows 125 seconds. If the operating system is rebooting, it may only have 5 seconds. If it is being shut down, it might have 20 seconds. But in these last cases, if the service runner has to be killed, then the child processes will also end up being killed, which is all the runner could do anyway at that point.

I picked 61 seconds because it's twice the default housekeeping interval, plus 1 second, and seems to be in a reasonable range.

Reply:
I think it would be valuable if this rationale is included in the comment so it is not lost in the pull request comments only.